### PR TITLE
content/blog: Add blog post for GSoC'24: Multiboot2

### PR DESCRIPTION
**Pull request by mariapana**
Added a blog post with the progress made during the first three weeks of GSoC'24.

Signed-off-by: Maria Pana <maria.pana4@gmail.com>

### CODE REVIEW

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+title: "Multiboot2 Support in Unikraft"
+description: |
+  Multiboot2, an enhanced version of Multiboot, solves some of the limitations of the original protocol, such as the lack of support for 64-bit systems. It also provides a more accurate and detailed communication during the boostrapping process and offers a standardized tag system for easier configuration.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:4]: support for 64-bit UEFI systems*

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+- Maria Pana
+---
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+## GSoC'24: Multiboot2 Support in Unikraft
+
+### Project overview
+
+When powering on an x86 system, it starts in Real Mode (the CPU is in 16-bit mode, and the memory is limited to 4MB).
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:20]: limited to 1MiB*
The address bus is 20 bits wide in Real Mode.

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+However, programming in Real Mode is limited, so the system needs to switch to Protected Mode (32-bit mode) or Long Mode (64-bit mode).
+This is where the bootloader comes in.
+After its initial routines, BIOS looks for bootable devices, such as a CD-ROM.
+If it finds one, the control is passed to the bootloader (a small portion of the executable code, found at the beginning of the device) which will then load the OS kernel into memory and transfer control to it.
+
+There are multiple bootloader standards and protocols, some already supported by Unikraft.
+Nonetheless, there is still room to add other bootloader standards, such as Multiboot2, which is the focus of this project.
+
+Multiboot2, an enhanced version of Multiboot, solves some of the limitations of the original protocol, such as the lack of support for 64-bit systems.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:31]: support for 64-bit UEFI systems*

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Making changes or improvements to software generally involves three phases: understanding the existing code, making the necessary changes, and testing the new code.
+This week I focused on the first phase, by revisting the [GNU Multiboot2 specification](https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html) and the notes I made prior to applying to GSoC.
+I also took a look at a previous attempt to integrate Multiboot2 in Unikraft and compared it to the original specification.
+
+Following the advice of my mentors, I then learned how to build GRUB and Multiboot2 images, and built GRUB from source.
+This helped me understand the process of creating a bootable image for a custom kernel and, by attaching GDB to the built GRUB, I prepared my "weapon" for future debugging sessions.
+
+Using the insight gained from these activities, I took a look at Unikraft's `mkukimg` script for creating bootable images, which will need updating to support Multiboot2.
+
+Lastly, I researched Unikraft's current Multiboot implementation I, as well as how other projects handled Multiboot2 integration to gain a wider perspective on the overall flow of the process moving forward.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:50]: > implementation I,

unnecessary `I,` here? :thinking:

**Review by mogasergiu — CHANGES_REQUESTED** [review]

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+---
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+## GSoC'24: Multiboot2 Support in Unikraft
+
+### Project overview
+
+When powering on an x86 system, it starts in Real Mode (the CPU is in 16-bit mode, and the memory is limited to 1MB).
+In order to gain more information about the system, it can interact with BIOS, which initially runs self test and initialization routines on hardware.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:21]: Hmm it is not clear who interacts with the BIOS and who wants to get more information about the system.
I would reword this sentence together with the last one to keep it clear :thinking:. I will try to shed some more light on things in the hopes that it will help you when rewording these :smiley:. I know the documentation for these is not always straightforward.

There exist two types of BIOS specifications (that I know of): PnP (Plug and Play) BIOS (aka legacy BIOS) and UEFI BIOS. Both are the same thing really, the difference comes to what they offer as interfaces to the software succeeding them (a bootloader for example) and the steps they take to achieve the state of being able to offer such interfaces (e.g. legacy BIOS interrupt routines vs UEFI System Table).

On x86 systems, regardless of whether PnP or UEFI, they start from a reset vector in Real Mode at a certain hardcoded address (I will let you find its value). From then on they can jump between Protected Mode, Unreal Mode, or 64-bit mode of Long Mode (UEFI 64-bit case) to achieve their goals.

PnP BIOS traditionally after starting in Real Mode, it may jump between Unreal Mode and Protected Mode to initialize hardware and set up resources that it can pass to the succeeding software to describe the systems to it in a manner that is motherboard/chipset/whatever agnostic (e.g. the memory map you saw in Multiboot2 actually originates from a BIOS call) only to then look through the list of bootable devices. The first one it finds that has a valid MBR, whose first few hundred bytes contain the bootstrap code is the one which control is passed to. But right before that, BIOS descends back to Real Mode (if it was not already there, it has full freedom to operate in whatever mode it wants to) as this is how it is agreed upon through the PnP specification, that the MBR bootstrap code starts in Real Mode (that code is also loaded at a very specific address, I will let you find out what that is).

UEFI BIOS similarly to the PnP one may jump between modes but may mostly run in 64-bit mode depending on the implementation (vendor's choice, but 32-bit UEFI is rare) and follows conceptually similar initialization steps, just under different form (think different looking C structs, but the same purpose). Unlike legacy BIOS, it looks for bootable devices with a GPT (or a (P)MBR for backwards compatibility) partitioning scheme, finds the EFI System Partition (FAT filesystem) in the GPT and boots the binary that has a special name (I will leave you to finding out what that name is). It does so while still being in 64-bit (or 32-bit for the 32-bit UEFI we don't care about).

Once loaded by the BIOS, the bootloader makes use of the aforementioned interfaces installed by the BIOS to be able to load the kernel and fill in the boot protocol structures, e.g.:
- the bootloader finds the memory map by using either the [E820 call](<http://www.uruk.org/orig-grub/mem64mb.html>) in the case of legacy BIOS, or [GetMemoryMap](<https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html#efi-boot-services-getmemorymap>) in the case of UEFI systems
- the bootloader is able to load the kernel from storage (e.g. disk) by using [INT 13 call](<https://en.wikipedia.org/wiki/INT_13H>) in the case of legacy BIOS, or [LoadFile](<https://uefi.org/specs/UEFI/2.10/13_Protocols_Media_Access.html#efi-load-file2-protocol-loadfile>) in the case of UEFI

Of course, the bootloader may not use these interfaces at all if it does not want to :smile:. In fact, it is a good thing for a bootloader to use as little BIOS calls as possible (firmware tends to be buggy) and have its own drivers ready that it can control (though while some are avoidable e.g. calls for disk, some are unavoidable, e.g. memory map fetching). For example, Windows Boot Manger is a famously very well written bootloader which is said to have around 7 UEFI calls only (that is very little!), choosing to rely mostly on its own drivers instead.

So the need for a bootloader does not necessarily come from Real Mode being difficult, as it is very easy to jump to 64-bit. Essentially, a BIOS is needed to initialize the very motherboard/vendor specific things only it itself needs to know only to then install some interfaces to make it easy for the software that comes afterwards to discover the resources of the system in a generic manner, so that it does not need to have a driver for every single motherboard out there. A bootloader like GRUB helps here as it can know the BIOS interfaces for the kernel, which rids the kernel of the need to speak any of these. That being said, the bootloader phase is for the comfortability of the kernel, but it can be skipped entirely if the kernel also knows these interfaces (e.g. see [EFI stub](<https://docs.kernel.org/admin-guide/efi-stub.html>))

**Review by mogasergiu — CHANGES_REQUESTED** [review]

File: content/blog/2024-06-17-gsoc-multiboot2.mdx (hunk 5 above)
Comment (mariapana) [reply content/blog/2024-06-17-gsoc-multiboot2.mdx:21]: The wording is indeed unclear, but I am glad and grateful for your insight! Thank you! I will be able to provide a much clearer view of the process now :smiley: 

I will push the modification, please let me know if I should go into more specifics with the explanation. Thanks again!

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+Making changes or improvements to software generally involves three phases: understanding the existing code, making the necessary changes, and testing the new code.
+This week I focused on the first phase, by revisting the [GNU Multiboot2 specification](https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html) and the notes I made prior to applying to GSoC.
+I also took a look at a previous attempt to integrate Multiboot2 in Unikraft and compared it to the original specification.
+
+Following the advice of my mentors, I then learned how to build GRUB and Multiboot2 images, and built GRUB from source.
+This helped me understand the process of creating a bootable image for a custom kernel and, by attaching GDB to the built GRUB, I prepared my "weapon" for future debugging sessions.
+
+Using the insight gained from these activities, I took a look at Unikraft's `mkukimg` script for creating bootable images, which will need updating to support Multiboot2.
+
+Lastly, I researched Unikraft's current Multiboot implementation, as well as how other projects handled Multiboot2 integration to gain a wider perspective on the overall flow of the process moving forward.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:53]: > how other projects handled Multiboot2 integration

Can you give some examples? :thinking: (also links to the projects)

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+### Next Steps
+
+Up next, I aim to get started on the actual implementation of Multiboot2 in Unikraft by adding the standard structures and files needed for the protocol.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:57]: Can you give some examples of what are the Multiboot2 tags and/or the parts of the structure that Unikraft will use (and how it will make use of them).

**Review by mogasergiu — CHANGES_REQUESTED** [review]
Just some more "getting fancy" comments :neckbeard:

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+---
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+## GSoC'24: Multiboot2 Support in Unikraft
+
+### Project overview
+
+When booting an x86 system, the CPU starts in 16-bit mode and the memory is limited to 1MB (in Real Mode).
+In order to find the first instruction to execute, the CPU jumps to a specific memory address known as a reset vector (e.g. physical address 0xFFFF0 for Intel 8086 processors).
```
[the remark this answers is not in the document]
Comment (mariapana) [reply content/blog/2024-06-17-gsoc-multiboot2.mdx:21]: Will do. The "e.g." stems from finding 0xFFFFF0 for Intel 80286 and 0xFFFFFFF0 for Intel 80386. Is the actual address relevant to the reset vector or is it just the "16 bytes under" idea? Thanks!

[the remark this answers is not in the document]
Comment (mogasergiu) [reply content/blog/2024-06-17-gsoc-multiboot2.mdx]: Huh :sweat_smile:
So, 80286 is the old 16-bit only CPU.
80386 is the first 32-bit CPU from the x86 family.
The answer is kind of combined. It can be either just `0xFFFF0` or `0xFFFFFFF0` as well. This has to do with what is known as BIOS shadowing.
When you press the Power On button, it is hardwired that the BIOS starts executing directly from flash that is at address `0xFFFF0` (Memory cycles in range 0xF0000 -> 0xFFFFF go to flash). This technique is called XIP (eXecute In Place), as RAM does not exist yet, the RAM controller has to be initialized, it's as if you would be executing code directly from storage. Modern chipsets however also decode memory cycles toward the last 2 MB of the frst 4GB (FFE0_0000 -> FFFF_FFFF) as memory cycles towards this flash memory (Memory aliasing). So basically *0xFFFF0 == *0xFFFFFFF0. As the BIOS initializes stuff, it enables 32-bit mode and may jump to execute from that top 4GB instead, but it's aliased so it executes the same code, it's just that `eip` instead of pointing to some low 16-bit address it points to some higher 32-bit address. At this stage, it enables RAM and then it programs some chipset registers known as PAM (Programmable Attribute Map) and here it has the option to allow memory cycles (either read, write or both, it can choose!) that go to 0xF0000 - 0xFFFFF to go to RAM instead (thus disabling aliasing). And this is how the BIOS ends up copying itself to RAM to execute faster, it will basically do a memcpy(0xF0000, 0xFFF00000, ...) and then jump again at that low 16-bit address which is now (faster) RAM instead of (slower) flash. 
As you can imagine, '286 being only 16-bit, does not go through this whole ordeal.
So, it should always be `0xffff0` regardless of generation. It's just that starting with 32-bit there is some aliasing going on so that BIOS is able to run faster at some point.

[the remark this answers is not in the document]
Comment (mogasergiu) [reply content/blog/2024-06-17-gsoc-multiboot2.mdx]: Huh, out of curiosity this morning I decided to recheck, because I vaguely remember having seen something among those lines in UEFI as well. And Wikipedia indeed says `0xFFFFFFF0` and some Stackoverflow answers [agree](<https://electronics.stackexchange.com/questions/509025/reason-for-x86-reset-vector-being-at-0xfffffff0-instead-of-0xffffffff>) and some [disagree](<https://stackoverflow.com/questions/25470971/hooking-x86-reset-vector#:~:text=The%20reset%20vector%20for%20any,in%20software%20to%20prevent%20it.>) :thinking:. So then I decided to really confirm by actually checking the manual again to make sure I am not misremembering. And, indeed, the chapte 10.1.4 of the third volume of Intel SDM's gives the answer:
```
The first instruction that is fetched and executed following a hardware reset is located at physical address
FFFFFFF0H. This address is 16 bytes below the processor’s uppermost physical address. The EPROM containing the
software-initialization code must be located at this address.
The address FFFFFFF0H is beyond the 1-MByte addressable range of the processor while in real-address mode. The
processor is initialized to this starting address as follows. The CS register has two parts: the visible segment
selector part and the hidden base address part. In real-address mode, the base address is normally formed by
shifting the 16-bit segment selector value 4 bits to the left to produce a 20-bit base address. However, during a
hardware reset, the segment selector in the CS register is loaded with F000H and the base address is loaded with
FFFF0000H. The starting address is thus formed by adding the base address to the value in the EIP register (that
is, FFFF0000 + FFF0H = FFFFFFF0H).
The first time the CS register is loaded with a new value after a hardware reset, the processor will follow the normal
rule for address translation in real-address mode (that is, [CS base address = CS segment selector * 16]). To
ensure that the base address in the CS register remains unchanged until the EPROM based software-initialization
code is completed, the code must not contain a far jump or far call or allow an interrupt to occur (which would
cause the CS selector value to be changed).
```
So it does indeed start in Unreal Mode, as far as addressing for instructions fetching goes, but the data segments are still very much 16-bit only. So then I checked the UEFI code. The [reset vector indeed starts 16-byte under 4G](<https://github.com/tianocore/edk2/blob/master/UefiCpuPkg/ResetVector/Vtf0/Ia16/ResetVectorVtf0.asm#L51>) and then flat addressing is set up for the [data segments as well](<https://github.com/tianocore/edk2/blob/master/UefiCpuPkg/ResetVector/Vtf0/Ia16/Real16ToFlat32.asm#L24>). I also rechecked the other chipset register stuff I mentioned (PAM) and that is indeed still true. So I have to correct myself here :nerd_face: :point_up:. The BIOS still does the same aforementioned steps, except that it starts directly at that higher address and when it sets up 32-bit, it only does so for data but not for code. Never make assumptions late at night based on what you think you remember :sweat_smile: always RTFM.

So, I guess, you can also add `0xFFFFFFF0` so that we can show off :rofl:.

[the remark this answers is not in the document]
Comment (mariapana) [reply content/blog/2024-06-17-gsoc-multiboot2.mdx]: Wow, thank you for the clearing up! I'll update the reset vector value to illustrate all possible values for the x86 family to avoid any confusion :smile:

**Comment by StefanJum** [comment]
@mariapana you need to squash the commits, so the PR contains only one commit in the end.
Also please mark the conversations as resolved when you address the comments so we don't get lost in them.

@mogasergiu can you please take another look content-wise, so we can merge and publish today?

**Comment by mariapana** [comment]
On it! :saluting_face:

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+When booting an x86 system, the CPU starts in 16-bit mode and the memory is limited to 1MB (in Real Mode).
+The CPU starts executing at a specific memory address known as a reset vector situated at 0xFFFF0 (for Intel 8086 processors) or 0xFFFFFFF0 (for Intel 80386 and later).
+The initial instructions found here are part of the BIOS, stored in flash memory on the motherboard.
+The BIOS's first goal is to run self test and initialization routines on hardware.
+During this phase, the BIOS sets up standardized interfaces (e.g. legacy BIOS interrupts or UEFI routines) that software that succeeds it (bootloader or operating system) will use to understand the computer's resources without needing special drivers for each different motherboard model.
+
+However, having the kernel directly deal with the interfaces installed by the BIOS unnecessarily increases complexity and maintenance effort.
+This is where the bootloader comes in.
+After its initial routines, BIOS looks for bootable devices, such as a CD-ROM.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:28]: > CD-ROM

Let's just say "such as a Hard Disk". There are several formats for CD-ROMs but I am not sure MBRs/GPTs are used for them and since this is what we give as an example on the next line, let's just go with "Hard Disk" or just "Disk" or SSD, whichever you want.

File: content/blog/2024-06-17-gsoc-multiboot2.mdx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+I also took a look at a previous attempt to integrate Multiboot2 in Unikraft and compared it to the original specification.
+
+Following the advice of my mentors, I then learned how to build GRUB and Multiboot2 images, and built GRUB from source.
+This helped me understand the process of creating a bootable image for a custom kernel and, by attaching GDB to the built GRUB, I prepared my "weapon" for future debugging sessions.
+
+Using the insight gained from these activities, I took a look at Unikraft's `mkukimg` script for creating bootable images, which will need updating to support Multiboot2.
+
+Lastly, I researched Unikraft's current Multiboot implementation, as well as how other projects handled Multiboot2 integration ([GRUB2](https://github.com/coreos/grub/tree/2.02-coreos/grub-core/loader), [Chromium OS](https://chromium.googlesource.com/chromiumos/third_party/grub2/+/fd01fd1d7f511093ee1c57e145ba25a254f9953e/loader), [ACRN Hypervisor](https://github.com/projectacrn/acrn-hypervisor/tree/53825c5cac898f682994d53d6ad0944c87fed6d6/hypervisor/boot)) to gain a wider perspective on the overall flow of the process moving forward.
```
Comment (mogasergiu) [inline content/blog/2024-06-17-gsoc-multiboot2.mdx:57]: > Chromium OS

Uh, I think this is a linux distro, so still GRUB :sweat_smile:

**Review by mogasergiu — CHANGES_REQUESTED** [review]
Two more nitpicks :sweat_smile:. I know I may seem to be excessively critical :rofl:. After this I will approve.

**Comment by mariapana** [comment]
I don't mind the critique at all, I am actually very grateful for your patience, guidance and effort to get the best out of the blog post :smile:. Thank you very much!!

**Review by mogasergiu — APPROVED** [review]
LGTM. Maybe also add a small commit message besides the commit subject so that it is aligned with the other commits in the repo.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

**Comment by mogasergiu** [comment]
Oops, wait, I think you may have misunderstood me. What I meant was the commit message should overall look smth like
```
blog: Add the first GSoC blog post about Multiboot2  <----------- commit subject

Add the first GSoC blog post about Multiboot2 offering an overview of   ]    commit
the project and the progress made during the first 3 weeks.             ]    message

Signed-off-by: Maria Pana <maria.pana4@gmail.com>
```
Notice that the way it is now, the subject line has "..." at the end, which is not aesthetically pleasing :point_up:.
Generally, you should try to have:
- max 70 chars for commit subject
- max 75 chars per commit message line

**Comment by mariapana** [comment]
Oh I see what you meant now :) I'm updating it now.

**Review by StefanJum — APPROVED** [review]
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>
